### PR TITLE
8314275: Incorrect stepping in switch

### PR DESCRIPTION
**Pull request by biboudis**
In code like the one below, the observable effect was that if the user debugs the following method, `multiply` called with `1`, the debugger steps from line 4 to line 6 (instead of 7) giving the impression that line 6 is going to be executed. The reason is that the `LineNumberTable` did not include an entry at the end of the `switch` mapping the statement that follows, to bytecode.

```java
1 private static double multiply(Integer i) { 
2   double cr = 15;                              
3    cr = switch (i) {                            
4        case 1 -> cr * 1;                       
5        case 2 -> cr * 2;                      
6        default -> cr * 4;                     
7    };                                           
8    return cr;                              
9 }   
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314275](https://bugs.openjdk.org/browse/JDK-8314275): Incorrect stepping in switch (**Bug** - P3)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17772/head:pull/17772` \
`$ git checkout pull/17772`

Update a local copy of the PR: \
`$ git checkout pull/17772` \
`$ git pull https://git.openjdk.org/jdk.git pull/17772/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17772`

View PR using the GUI difftool: \
`$ git pr show -t 17772`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17772.diff">https://git.openjdk.org/jdk/pull/17772.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17772#issuecomment-1934252641)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java [hunk 1]
```diff
@@ -1428,6 +1428,9 @@ private void handleSwitch(JCTree swtch, JCExpression selector, List<JCCase> case
                     code.put4(caseidx + 4, offsets[i]);
                 }
             }
+
+            // Emit line position for the end of the switch
+            code.statBegin(TreeInfo.endPos(swtch));
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:1433]: shouldn't this change affect already existing tests like: RuleSwitchBreaks.java in the same folder: test/langtools/tools/javac/classfiles/attributes/LineNumberTable/?

**Review by vicente-romero-oracle — APPROVED** [review]
looks good, just a minor comment

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:1433]: no I was just wondering why the issue didn't appear in the previous test, but what you are saying could be the explanation

Comment (biboudis) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java]: What do you think about this? https://github.com/openjdk/jdk/pull/17772/commits/a836d976e56c624882ae793ee89b1146c9fd4352

(also notice the newly added test case. For the `switch` statement the line 8 is not included and for the `switch` expression it is.)

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java]: yep looks good, thanks

**Pull request closed without merging**